pandas/core/_numba/extensions.py

```diff
@@ -0,0 +1,426 @@
+"""
+Utility classes/functions to let numba recognize
+pandas Index/Series/DataFrame
+
+Mostly vendored from https://github.com/numba/numba/blob/main/numba/tests/pdlike_usecase.py
+"""
+
+from __future__ import annotations
+
+import numba
+from numba.core import (
+    cgutils,
+    types,
+)
+from numba.core.datamodel import models
+from numba.core.extending import (
+    NativeValue,
+    box,
+    lower_builtin,
+    make_attribute_wrapper,
+    register_model,
+    type_callable,
+    typeof_impl,
+    unbox,
+)
+from numba.core.imputils import impl_ret_borrowed
+
+from pandas.core.frame import DataFrame
+from pandas.core.indexes.base import Index
+from pandas.core.series import Series
+
+
+# TODO: Range index support
+# (not passing an index to series constructor doesn't work)
+class IndexType(types.Buffer):
+    """
+    The type class for Index objects.
+    """
+
+    array_priority = 1000
+
+    def __init__(self, dtype, layout, pyclass) -> None:
+        self.pyclass = pyclass
+        super().__init__(dtype, 1, layout)
+
+    @property
+    def key(self):
+        return self.pyclass, self.dtype, self.layout
+
+    @property
+    def as_array(self):
+        return types.Array(self.dtype, 1, self.layout)
+
+    def copy(self, dtype=None, ndim: int = 1, layout=None):
+        assert ndim == 1
+        if dtype is None:
+            dtype = self.dtype
+        layout = layout or self.layout
+        return type(self)(dtype, layout, self.pyclass)
+
+
+class SeriesType(types.ArrayCompatible):
+    """
+    The type class for Series objects.
+    """
+
+    array_priority = 1000
+
+    def __init__(self, dtype, index) -> None:
+        assert isinstance(index, IndexType)
+        self.dtype = dtype
+        self.index = index
+        self.values = types.Array(self.dtype, 1, "C")
```

Comment:
can this be something more distinctive than "values"?  we use that to mean a bunch of different things around the codebase